src/daemons/publishDaemon.ts

```diff
@@ -0,0 +1,120 @@
+import { AMQPMessage } from '@cloudamqp/amqp-client'
+import admin from 'firebase-admin'
+
+import { getApiKeyByKey } from '../db/couchApiKeys'
+import { getDeviceById, getDevicesByLoginId } from '../db/couchDevices'
+import { DbConnections } from '../db/dbConnections'
+import { logger } from '../util/logger'
+import { asRabbitMessage, SendableMessage } from '../util/pushSender'
+import { runDaemon } from './runDaemon'
+
+runDaemon(async tools => {
+  const { connections } = tools
+
+  const consumer = await connections.queue.subscribe(
+    { noAck: false },
+    message => {
+      handleMessage(connections, message).catch(error => {
+        logger.warn({
+          msg: 'Failed to handle message',
+          body: message.bodyToString(),
+          error
+        })
+      })
+    }
+  )
+
+  // Will wait until we have an error, which closes the channel:
+  await consumer.wait()
+})
+
+async function handleMessage(
+  connections: DbConnections,
+  amqpMessage: AMQPMessage
+): Promise<void> {
+  const { deviceId, loginId, message } = asRabbitMessage(
+    amqpMessage.bodyToString() ?? '{}'
+  )
+
+  // Send it where it needs to go:
+  if (deviceId != null) {
+    await sendToDevice(connections, deviceId, message)
+  }
+  if (loginId != null) {
+    const deviceRows = await getDevicesByLoginId(connections, loginId)
+    for (const deviceRow of deviceRows) {
+      await sendToDevice(connections, deviceRow.device.deviceId, message)
+    }
```

Comment:
I expected that batching would be implemented. A maximum of 500 messages can be batched over the Firebase API. This significantly improves throughput AFAIU.